README
IRC bot


Changes
Revision history for Perl extension App::Mobijiro

0.02    Thu Apr 28 23:59:18 2011
        - use cpan module style

0.01    Mon Apr 25 19:26:07 2011
        - original version


mobijiro.pl
use common::sense;
use Encode;
use Socket;

use AnyEvent;
use AnyEvent::IRC::Client;
use Tatsumaki::HTTPClient;

use URI;
use Web::Scraper;
use Data::Validate::URI ();

our $CONFIG;

if (-f 'settings.pl') {
    $CONFIG = do 'settings.pl' or die $!;
} else {
    $CONFIG = {
        ch => '#',
        server => 'irc.freenode.net',
        port => 6667,
        info => {
            nick => '',
            user => '',
            real => 'the bot',
        },
    };
}

$CONFIG->{loopback} = inet_ntoa( inet_aton('localhost') );

#-----

local $| = 1;

our $CONNECTION = 0;
our $LT = scalar localtime;

my @conf = ($CONFIG->{server}, $CONFIG->{port}, $CONFIG->{info});

my $cv = AE::cv;

my $ua = Tatsumaki::HTTPClient->new;

my $cl;

my $ltw; $ltw = AE::timer 1, 1, sub {
    $LT = scalar localtime;
};

my $t; $t = AE::timer 5, 30, sub {
    say "[ $LT ] connection status : $CONNECTION";
    connect_to_server() unless $CONNECTION;
};

my $scraper = scraper {
    process '/html/head/title', 'title' => 'TEXT';
};

$cv->recv;

sub connect_to_server {
    say "[ $LT ] start connection";

    $cl = AnyEvent::IRC::Client->new;

    $cl->reg_cb(
        'connect' => sub {
            my ($cl, $err) = @_;

            if (defined $err) {
                print "[ $LT ] Connect ERROR! => $err\n";
                $CONNECTION = 0;
                $cv->broadcast;

            } else {
                print "[ $LT ] Connected! Yay!\n";
                $CONNECTION = 1;
            }
        },

        registered => sub {
            my ($self) = @_;

            print "[ $LT ] registered!\n";
            $cl->enable_ping(60);

            $cl->send_srv("JOIN", $CONFIG->{ch});
            $cl->send_chan($CONFIG->{ch}, "NOTICE", $CONFIG->{ch}, "hi, i'm a bot!");
        },

        irc_001 => sub { say "[ $LT ] irc_001"; },

        irc_privmsg => sub {
            my ($self, $msg) = @_;

            my $message = decode_utf8 $msg->{params}[1];

            my @url_list = $message =~ m{(https?://[\S]+)}g;

            process_url($_) for (@url_list);
        },

        disconnect => sub {
            $CONNECTION = 0;
            print "[ $LT ] Oh, got a disconnect: $_[1], exiting...\n";
        }
    );

    $cl->connect(@conf);
}

sub process_url {
    my $url = shift;

    say "[ $LT ] $url";

    return unless Data::Validate::URI::is_uri($url);

    if (is_twitter($url)) {
        send_twitter_status($url);
        return;
    }

    $ua->head($url, timeout => 3, sub {
        my $res = shift;

        my $remote = URI->new( $res->header('url') )->host;
        my $remote_addr = inet_ntoa( inet_aton($remote) );
        my $content_length = $res->headers->content_length ? $res->headers->content_length : 0;

        if ($remote_addr eq $CONFIG->{loopback}) {
            my $msg = sprintf "%s is loopback!", $url;
            $cl->send_chan($CONFIG->{ch}, "NOTICE", $CONFIG->{ch}, encode_utf8($msg));

        } elsif ($content_length > 0 && $content_length > 1024 * 1024) {
            my $msg = sprintf "Too large to fetch: %s [ %s ]", $url, $res->content_type;
            $cl->send_chan($CONFIG->{ch}, "NOTICE", $CONFIG->{ch}, encode_utf8($msg));

        } elsif ($res->headers->content_type ne 'text/html') {
            my $msg = sprintf "%s [%s]", $url, $res->headers->content_type;
            $cl->send_chan($CONFIG->{ch}, "NOTICE", $CONFIG->{ch}, encode_utf8($msg));

        } else {
            $ua->get($url, timeout => 3, sub {
                my $res = shift;

                my $info = {};
                if ($res->is_success) {
                    my $data = $scraper->scrape($res->decoded_content);
                    $info->{title} = $data->{title};
                    $info->{content_type} = $res->headers->content_type;

                } else {
                    $info->{title} = $res->status_line;
                }

                my $msg = sprintf "%s [%s] %s", $info->{title}, $info->{content_type}, $url;
                $cl->send_chan($CONFIG->{ch}, "NOTICE", $CONFIG->{ch}, encode_utf8($msg));
            });

        }

    });
}

sub is_twitter {
    return ($_[0] =~ m{^https?://twitter.com/(?:#!/)?[^/]+/status/\d+}) ? 1 : undef;
}

sub send_twitter_status {
    my $url = shift;

    $url =~ s{\.com/#!/}{.com/};

    $ua->get($url, timeout => 3, sub {
        my $res = shift;

        return unless ($res->is_success); 

        my $ts = scraper {
            process '.entry-content', 'tweet' => 'TEXT';
            process '.screen-name', 'screen_name' => 'TEXT';
        };

        my $status = $ts->scrape($res->decoded_content);

        my $msg = sprintf "<%s> %s / via %s", $status->{screen_name}, $status->{tweet}, $url;
        $cl->send_chan($CONFIG->{ch}, "NOTICE", $CONFIG->{ch}, encode_utf8($msg));
    });
}

__END__



lib/App/Mobijiro.pm
package App::Mobijiro;
use practical;
our $VERSION = '0.02';
use constant DEBUG => $ENV{MOBIJIRO_DEBUG};
use parent 'Class::Accessor::Fast';
__PACKAGE__->mk_accessors(qw/
    loopback channel
/);

use AE;
use Encode;
use Socket;
use AnyEvent::IRC::Client;
use Tatsumaki::HTTPClient;
use URI;
use Web::Scraper;
use Data::Validate::URI;

our $CONNECTION = 0;

sub new {
    my $class = shift;
    my %args  = @_;

    return bless {
        %args,
        ua       => Tatsumaki::HTTPClient->new,
        client   => AnyEvent::IRC::Client->new,
        loopback => inet_ntoa( inet_aton('localhost') ),
    }, $class;
}

sub run {
    my $self = shift;
    my $con_watcher; $con_watcher = AE::timer 5, 30, sub {
        say "con_watcher" if DEBUG;
        $self->connect unless $CONNECTION;
        undef $con_watcher;
    };
    return $con_watcher;
}

sub connect {
    my $self = shift;

    my %opts = (
        connect => sub {
            my ($cl, $err) = @_;

            if (defined $err) {
                say "Connect error: $err" if DEBUG;
                $CONNECTION = 0;
            } else {
                say "Connected" if DEBUG;
                $CONNECTION = 1;
            }
        },
        registered => sub {
            say "Registered" if DEBUG;
            $self->{client}->enable_ping(60);
            $self->{client}->send_srv("JOIN", $self->channel);
            $self->{client}->send_chan($self->channel, "NOTICE", $self->channel, "Hi, I'm a bot!");
        },
        irc_privmsg => sub {
            my $msg = Encode::decode_utf8($_[1]->{params}[1]);
            my @url = $msg =~ m{(https?://[\S]+)}g;
            $self->resolve($_) for @url;
        },
        disconnect => sub {
            $CONNECTION = 0;
            say "Disconnected: $_[1]" if DEBUG;
        },
    );

    $self->{client}->reg_cb(%opts);
    $self->{client}->connect($self->{server}, $self->{port}, $self->{info})
}

sub resolve {
    my ($self, $url) = @_;

    return unless Data::Validate::URI::is_uri($url);

    $self->{ua}->head($url, sub {
        my $res = shift;
        my $h = $res->headers;
        say "HEAD: $url" if DEBUG;

        # NOTE: $host must be predecleared for $remote->{addr}
        my $host = URI->new( $h->header('url') )->host;

        my $remote = {
            host   => $host,
            addr   => inet_ntoa( inet_aton($host) ),
            length => $h->content_length ? $h->content_length : 0,
            url    => $h->header('url')  ? $h->header('url')  : $url,
            type   => $h->content_type   ? scalar $h->content_type : 'text/plain',
                # NOTE: content_type should be called in scalar context
        };

        if ($remote->{addr} eq $self->loopback) {
            $self->send(sprintf "Loopback: %s", $url);
            return;
        }

        if ($remote->{length} > 1024 * 1024) {
            $self->send(sprintf "Large: %s [%s]", $url, $res->content_type);
            return;
        }

        if ($remote->{type} ne 'text/html') {
            $self->send(sprintf "%s [%s]", $url, $remote->{type});
            return;
        }

        if ($self->is_twitter($remote->{url})) {
            $remote->{url} =~ s{\.com/#!/}{.com/};
            say "Twitter: $remote->{url}" if DEBUG;
            $self->{ua}->get($remote->{url}, sub {
                my $res = shift;

                unless ($res->is_success) {
                    $self->send(sprintf "Error: %s", $res->status_line);
                    return;
                }

                my $ts = scraper {
                    process '.entry-content', 'tweet' => 'TEXT';
                    process '.screen-name', 'name' => 'TEXT';
                };

                my $s = $ts->scrape($res->decoded_content);

                $self->send(sprintf "<%s> %s / via %s", $s->{name}, $s->{tweet}, $remote->{url});
            });
            return;
        }

        $self->{ua}->get($url, sub {
            my $res = shift;
            my $h = $res->headers;
            say "GET: $url" if DEBUG;

            my $info = {};
            if ($res->is_success) {
                my $scraper = scraper {
                    process '/html/head/title', 'title' => 'TEXT';
                };
                my $data = $scraper->scrape($res->decoded_content);
                $info->{title} = $data->{title};
                $info->{type} = $h->content_type;
            } else {
                $info->{title} = $res->status_line;
                $info->{type} = '';
            }

            $self->send(sprintf "%s [%s] %s", $info->{title}, $info->{type}, $h->header('url'));
        });
    });
}

sub send {
    my $self = shift;
    $self->{client}->send_chan(
        $self->channel, "NOTICE", $self->channel, Encode::encode_utf8(shift)
    );
}

sub is_twitter {
    my $self = shift;
    say "Is Twitter: $_[0]" if DEBUG;
    return ($_[0] =~ m{^https?://twitter.com/(?:#!/)?[^/]+/status/\d+}) ? 1 : undef;
}


1;
__END__

=head1 NAME

App::Mobijiro -

=head1 SYNOPSIS

  use App::Mobijiro;

=head1 DESCRIPTION

App::Mobijiro is

=head1 AUTHOR

punytan E<lt>punytan@gmail.comE<gt>

=head1 SEE ALSO

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
